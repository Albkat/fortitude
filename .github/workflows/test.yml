name: Tests

on:
  push:
    paths:
      - '**.rs'
  pull_request:
    paths:
      - '**.rs'

env:
  CARGO_TERM_COLOR: always

jobs:

  build:
    name: Build and Test
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Build
      run: cargo build --verbose
    - name: Run tests
      run: cargo test --all-targets --all-features --verbose

  cargo-install:
    name: Test cargo installation
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Check cargo install
      run: |
        cargo install --path fortitude
        fortitude --help

  pip-install:
    name: Test pip installation
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Set up Python
      uses: actions/setup-python@v5
      with:
        python-version: '3.x'
    - name: Check pip install
      run: |
        pip install .
        fortitude --help
