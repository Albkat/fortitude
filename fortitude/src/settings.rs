use ruff_macros::CacheKey;
/// A collection of user-modifiable settings. Should be expanded as new features are added.
use std::fmt::{Display, Formatter};

use crate::rule_selector::RuleSelector;

pub struct Settings {
    pub line_length: usize,
}

impl Default for Settings {
    fn default() -> Self {
        Self { line_length: 100 }
    }
}

/// Toggle for rules still in preview
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, CacheKey, is_macro::Is)]
pub enum PreviewMode {
    #[default]
    Disabled,
    Enabled,
}

impl From<bool> for PreviewMode {
    fn from(version: bool) -> Self {
        if version {
            PreviewMode::Enabled
        } else {
            PreviewMode::Disabled
        }
    }
}

impl Display for PreviewMode {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Disabled => write!(f, "disabled"),
            Self::Enabled => write!(f, "enabled"),
        }
    }
}

/// Default rule selection
pub const DEFAULT_SELECTORS: &[RuleSelector] = &[RuleSelector::All];
