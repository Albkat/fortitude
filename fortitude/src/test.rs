use std::path::Path;

use anyhow::Result;
use itertools::join;
use ruff_source_file::{SourceFile, SourceFileBuilder};

use crate::{
    check::{
        ast_entrypoint_map, check_file, read_to_string, rules_to_path_rules, rules_to_text_rules,
    },
    message::DiagnosticMessage,
    rules::Rule,
    settings::Settings,
};

pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a Fortran file in the `resources/test/fixtures` directory.
pub(crate) fn test_path(
    path: impl AsRef<Path>,
    rules: &[Rule],
    settings: &Settings,
) -> Result<String> {
    let path = test_resource_path("fixtures").join(path);
    let source = read_to_string(&path)?;
    let filename = path.to_string_lossy();
    let file = SourceFileBuilder::new(filename.as_ref(), source.as_str()).finish();
    Ok(test_contents(&path, &file, rules, settings))
}

pub(crate) fn test_contents(
    path: &Path,
    file: &SourceFile,
    rules: &[Rule],
    settings: &Settings,
) -> String {
    let path_rules = rules_to_path_rules(rules);
    let text_rules = rules_to_text_rules(rules);
    let ast_entrypoints = ast_entrypoint_map(rules);

    match check_file(
        &path_rules,
        &text_rules,
        &ast_entrypoints,
        path,
        file,
        settings,
    ) {
        Ok(violations) => {
            if violations.is_empty() {
                return String::new();
            }
            let strings = violations
                .into_iter()
                .map(|v| DiagnosticMessage::from_ruff(file, v))
                .map(|d| d.to_string());
            join(strings, "\n")
        }
        Err(msg) => {
            panic!("Failed to process: {msg}");
        }
    }
}
