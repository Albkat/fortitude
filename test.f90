! This function should trigger for missing enclosing module
integer function double(x)
  integer, intent(in) :: x
  double = 2 * x
end function

subroutine print_val(y)
  real, intent(in) :: y
  write (*,*) y
end subroutine print_val

module my_module

  ! Should not raise
  use, intrinsic :: iso_fortran_env, only: dp => real64

  ! Should raise for lack of only clause
  use, intrinsic :: iso_c_binding

  implicit none
  private

  ! Should not raise error
  real(dp), parameter :: pi = 3.14159265358979_dp

  complex(dp), parameter :: imag = (0.0_8, 1.0_8)

  ! Should raise errors for missing suffix
  real(dp), parameter :: pi_32 = 3.14159265358979
  real(dp), parameter :: pi_short = 3.1415

  ! Should raise syntax error
  real(dp), parameter :: mistake = 2e

  ! Should not raise error for maximum line length
  character(*), parameter :: long_string = "https://verylongurl.com/page/another_page/yet_another_page/wow"

  ! Should report error
  !int, parameter :: bad_type = 1

  ! Should not raise error for maximum line length
  character(*), parameter :: long_comment = "short string" ! The origins of this string date back to the 18th century...

  ! Should raise error for non-standard 'type*N'
  logical*4, parameter :: true = .true.

  ! Should raise error for non-standard length
  character*20 :: foo

  ! TODO should raise error for outdated 'character*(*)'
  character*(*), parameter :: hello = "hello world"

  interface
    ! This function should raise an error for missing implicit none, one for using
    ! a number literal kind in the signature, and one for a number literal kind in the
    ! variable list.
    integer(8) function interface_func(x)
      integer(kind=8), intent(in) :: x
    end function

    ! This function shouldn't raise anything
    real function interface_func2(x)
      implicit none
      real, intent(in) :: x
    end function
  end interface

contains

  ! Should not trigger linter
  integer function double(x)
    integer, intent(in) :: x
    double = 2 * x
  end function

  ! Should not trigger linter
  subroutine triple(x)
    integer, intent(inout) :: x
    x = x * 3
  end subroutine

  ! Should trigger for superfluous implicit none
  integer function quad(x)
    implicit none
    integer, intent(in) :: x
    quad = 4 * x
  end function

  ! Should trigger for use of 'double precision'
  double precision function double_prec(x)
    double precision, intent(in) :: x
    double_prec = 2 * x
  end function
end module

! This function should trigger for missing enclosing module
subroutine triple(x)
  integer, intent(inout) :: x
  x = x * 3
end subroutine

! Should trigger for missing implicit none
module implicit_module
  implicit integer(A)
  parameter(N = 1)
end module

! Should trigger for missing implicit none
program myprog
  write(*,*) "Hello world!"
end program
