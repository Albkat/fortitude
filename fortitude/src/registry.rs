use fortitude_macros::RuleNamespace;

use crate::rules::Rule;

// Rule categories and identity codes
// ----------------------------------
// Helps to sort rules into logical categories, and acts as a unique tag with which
// users can switch rules on and off.

pub trait AsRule {
    fn rule(&self) -> Rule;
}

impl Rule {
    pub fn from_code(code: &str) -> Result<Self, FromCodeError> {
        // TODO(peter): second var and lhs should be `code`
        let (category, _) = Category::parse_code(code).ok_or(FromCodeError::Unknown)?;
        category
            .all_rules()
            .find(|rule| rule.noqa_code().suffix() == code)
            .ok_or(FromCodeError::Unknown)
    }
}

#[derive(thiserror::Error, Debug)]
pub enum FromCodeError {
    #[error("unknown rule code")]
    Unknown,
}

pub enum RuleCheckKind {
    Text,
    Path,
    AST,
}

/// The category of each rule defines the sort of problem it intends to solve.
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, RuleNamespace)]
pub enum Category {
    /// Failure to parse a file.
    #[prefix = "E"]
    Error,
    /// Violation of style conventions.
    #[prefix = "S"]
    Style,
    /// Misuse of types and kinds.
    #[prefix = "T"]
    Typing,
    /// Failure to use modules or use them appropriately.
    #[prefix = "M"]
    Modules,
    /// Best practices for setting floating point precision.
    #[prefix = "P"]
    Precision,
    /// Check path names, directory structures, etc.
    #[prefix = "F"]
    Filesystem,
}

pub trait RuleNamespace: Sized {
    /// Returns the prefix that every single code that fortitude uses to identify
    /// rules from this category starts with.
    fn common_prefix(&self) -> &'static str;

    /// Attempts to parse the given rule code. If the prefix is recognized
    /// returns the respective variant along with the code with the common
    /// prefix stripped.
    fn parse_code(code: &str) -> Option<(Self, &str)>;

    #[allow(dead_code)]
    fn name(&self) -> &'static str;

    #[allow(dead_code)]
    fn description(&self) -> &'static str;
}
