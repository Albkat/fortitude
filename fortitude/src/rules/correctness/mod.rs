pub mod implicit_kinds;
pub mod kind_suffixes;
pub mod select_default;
pub mod trailing_backslash;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::settings::Settings;
    use crate::test::test_path;

    #[test_case(Rule::MissingDefaultCase, Path::new("C001.f90"))]
    #[test_case(Rule::TrailingBackslash, Path::new("C011.F90"))]
    #[test_case(Rule::NoRealSuffix, Path::new("C021.f90"))]
    #[test_case(Rule::ImplicitRealKind, Path::new("C022.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
