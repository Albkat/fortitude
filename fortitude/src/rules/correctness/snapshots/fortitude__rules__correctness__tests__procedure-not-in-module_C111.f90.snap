---
source: fortitude/src/rules/correctness/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/correctness/C111.f90:1:1: C111 function not contained within (sub)module or program
  |
1 | integer function double(x)
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^ C111
2 |   integer, intent(in) :: x
3 |   double = 2 * x
  |

./resources/test/fixtures/correctness/C111.f90:6:1: C111 subroutine not contained within (sub)module or program
  |
4 | end function double
5 |
6 | subroutine triple(x)
  | ^^^^^^^^^^^^^^^^^^^^ C111
7 |   integer, intent(inout) :: x
8 |   x = 3 * x
  |
